src/Migration.php

```diff
@@ -28,10 +28,14 @@ class Migration {
 	 */
 	public function run_migrations() {
 		$current_db_version = get_option( Options::WC_BLOCK_VERSION, '' );
+		$schema_version     = get_option( 'wc_blocks_db_schema_version', '' );
 
 		// This check is necessary because the version was not being set in the database until 10.3.0.
+		// Checking wc_blocks_db_schema_version determines if it's a fresh install (value will be empty)
```

Comment:
I would suggest exploring an alternative route here, as the assumption that `wc_blocks_db_schema_version` is empty on fresh installs is a bit fragile. This option is updated whenever [`maybe_create_tables`](https://github.com/woocommerce/woocommerce-blocks/blob/e59611bbccda1cf5133d64f787fc8d92d9161b06/src/Installer.php#L73) is triggered and this method is hooked to `'admin_init'`, which is fired whenever an admin screen or script is being initialized, so fresh installs will have the default 0 value updated to 260 whenever the table creation is complete.

Reply:
This code will be executed before that `admin_init`  is triggered. Empty returns true if you pass 0 or "".

Reply:
> This code will be executed before that admin_init  is triggered. Empty returns true if you pass 0 or "".

This option is not deleted from the DB when the plugin is removed, so if someone at some point in the past had Woo and/or Woo blocks up and running, removed them, and now enabled them again, `$this->wc_blocks_update_1030_blockified_product_grid_block();` will be triggered and cause the same problem.

> I agree that checking wc_blocks_db_schema_version isn't the best approach, but unfortunately, it seems that it is the only way to check if it is a new installation.

Just to clarify, is the goal here to run the migration to all users that already had Woo installed while at the same time displaying the blockified template for brand-new users?

Reply:
> > This code will be executed before that admin_init  is triggered. Empty returns true if you pass 0 or "".
> 
> This option is not deleted from the DB when the plugin is removed, so if someone at some point in the past had Woo and/or Woo blocks up and running, removed them, and now enabled them again, `$this->wc_blocks_update_1030_blockified_product_grid_block();` will be triggered and cause the same problem.
> 

I think that it is fine since it is an uncommon use case. The user had an old version of WC Core. Remove it. Install the new one. The user will have the classic templates. 

> > I agree that checking wc_blocks_db_schema_version isn't the best approach, but unfortunately, it seems that it is the only way to check if it is a new installation.
> 
> Just to clarify, is the goal here to run the migration to all users that already had Woo installed while at the same time displaying the blockified template for brand-new users?

Yes.

Reply:
This is a good idea, but at the same time, if the user updates from WooCommerce Core 7.8 to WooCommerce Core 7.9 and doesn't have any products, we want to keep it using the classic template. This is the migration strategy that we have communicated and for this reason, I would not change the logic.

> This way we have something more robust to rely on instead of the value of a DB migration to determine if the blockified template should be displayed or not.

Not sure why you are saying this. We have full control of when the value is inserted in the DB, so I don't see any particular issue.

Reply:
regarding: 

> I think that it is fine since it is an uncommon use case. The user has an old version of WC Core. Remove it. Install the new one. The user will have the classic templates.

and:

> Not sure why you are saying this. We have full control of when the value is inserted in the DB, so I don't see any particular issue.

This migration is triggered on `'plugins_loaded'` (which is fired once activated plugins have loaded), so if someone installed the plugin on a new store, did some changes, deactivated the plugin and activated again, the value of `wc_blocks_db_schema_version` might not be 0 anymore (as all it takes to update this is accessing the WP admin page, firing the `'admin_init'` hook). One more example of how fragile this affirmation is:

> Checking wc_blocks_db_schema_version determines if it's a fresh install (value will be empty)

I don't agree with relying on this option for this and the aforementioned reasons.

> if the user updates from WooCommerce Core 7.8 to WooCommerce Core 7.9 and doesn't have any products, we want to keep it using the classic template

why?

Reply:
> This migration is triggered on `'plugins_loaded'` (which is fired once activated plugins have loaded), so if someone installed the plugin on a new store, did some changes, deactivated the plugin and activated again, the value of `wc_blocks_db_schema_version` might not be 0 anymore (as all it takes to update this is accessing the WP admin page, firing the `'admin_init'` hook). One more example of how fragile this affirmation is:

True, but this is the entire condition: 

```
if ( ! empty( $schema_version ) && ( empty( $current_db_version ) ) ) {
```

So, there is an additional check.

> why?

We worked on the migration strategy one year ago when the "blockified templates" were very far from ready. I think that we didn't think about :( 

I'm open to refactoring this, but since the migration logic has already been deployed and no issue has been reported, I would avoid this change and fix the current logic. Furthermore, in the following months, we will remove [it](https://github.com/woocommerce/woocommerce-blocks/pull/10415/files#diff-7a149914e4cec2dfd6fd142ebb3653fb186d871e694aea733ed1a14d1ec0b7f9R37-R40)

cc @kmanijak thoughts?

Reply:
> We worked on the migration strategy one year ago when the "blockified templates" were very far from ready. I think that we didn't think about :(

Thanks for the explanation. Regarding:

> if the user updates from WooCommerce Core 7.8 to WooCommerce Core 7.9 and doesn't have any products, we want to keep it using the classic template

This PR is planned to be released as part of version 10.6.3, right? Any changes made here shouldn't impact the update from version 7.8 to 7.9, correct?

Reply:
> This PR is planned to be released as part of version 10.6.3, right? Any changes made here shouldn't impact the update from version 7.8 to 7.9, correct?

Right, but it will impact the update from version 7.8 to 8.0.

Reply:
> Right, but it will impact the update from version 7.8 to 8.0.

Do you mind sharing an example of a real problem a user would have by updating from version 7.8 to 8.0 and allowing them to see the blockified template right away if they are a new store? I saw your explanation over here:

> We worked on the migration strategy one year ago when the "blockified templates" were very far from ready. I think that we didn't think about :(

but to be honest it is still not clear why we still want to display the classic template for such cases?

Reply:
>  it is still not clear why we still want to display the classic template for such cases?

The rule that was followed since the blockified templates were implemented (couple of months and versions ago) was:
- use blockified template in new stores
- respect the current setup of the existing stores that update the version

Considering there's time pressure to include that change in WC Core, I wouldn't start changing that logic at this very moment and include the fix.

---

Regarding the discussion above I would consider adding another criterion, except for the number of products or orders in store. If we were to go that way, I think we would have to also check if the Product Archive templates were edited. There may be other cases too. I'm not against refactoring or changing the rules, but for now, I think sticking with the rules above (and then the solution in this PR) keeps it simple and covers the case that we found broken.

Reply:
> The rule that was followed since the blockified templates were implemented (couple of months and versions ago) was:

> use blockified template in new stores
> respect the current setup of the existing stores that update the version

yep! But @gigitux is saying that there's an exception to this rule and that we should display the classic template even for new stores when updating from version 7.8 to 7.9 and from 7.8 to 8.0, that's the part that no explanation was provided at all.

There are many technical limitations with the solution provided here, and I don't feel comfortable with approving it for this reason, so I'm withdrawing my review.

Reply:
> yep! But @gigitux is saying that there's an exception to this rule and that we should display the classic template even for new stores when updating from version 7.8 to 7.9 and from 7.8 to 8.0, that's the part that no explanation was provided at all.

If you upgrade the version even if you don't have products, it means that the store isn't new:

![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/9fbb8bc8-d109-4264-b698-1d03e56a49c2)

Reply:
> If you upgrade the version even if you don't have products, it means that the store isn't new:

This graphic is just a representation of the current logic & unfortunately, doesn't answer my question: nothing in it explains why the plugin version is being used to determine if a store is new or not. And to reinforce my comment over here: https://github.com/woocommerce/woocommerce-blocks/pull/10415#issuecomment-1661825951 there's no proposal to change what is represented in this image.

Reply:
We started saving the version of WC Blocks in the database only with WooCommerce Blocks 10.3.0 release.

So, if this condition below is true, it means that it isn't a new installation because:
- $schema_version is already set
- $current_db_version is empty, so it means that the user is upgrading from a version of WC Blocks < 10.3.0
```
! empty( $schema_version ) && ( empty( $current_db_version )
```

Reply:
> nothing in it explains why the plugin version is being used to determine if a store is new or not.

I think "fresh WooCommerce installation" tries to answer it. It may be the insufficient choice of words, but this diagram was supposed to represent this table from [this PR](https://github.com/woocommerce/woocommerce-blocks/pull/6538):

![image](https://github.com/woocommerce/woocommerce-blocks/assets/20098064/947b7df4-ad9e-4f04-b637-c8973e031ffd)

which clearly differentiates "New installation" from "Upgrade" and that's what the first condition in the diagram was supposed to represent ("Fresh installation" meaning "New installation" and not "Upgrade"). And we'd like to preserve that logic. Does that help answer the concern?

Reply:
> And we'd like to preserve that logic. Does that help answer the concern?

Thanks @kmanijak , it does represent what the current implementation is, but unfortunately, I couldn't find any explanation why this decision was made on that PR =/. 

Either way, we still have all the aforementioned problems to clear out, as the solution provided here does not apply to all use-case scenarios, with very important limitations as flagged on https://github.com/woocommerce/woocommerce-blocks/pull/10415#discussion_r1280372021 , https://github.com/woocommerce/woocommerce-blocks/pull/10415#discussion_r1280425734 and  https://github.com/woocommerce/woocommerce-blocks/pull/10415#discussion_r1280656357 .

Since the decision was to proceed with this work for the patch release, I'll propose circling back to solve these problems in our next cycle.

Reply:
> but unfortunately, I couldn't find any explanation why this decision was made on that PR =/.

Unfortunately, I cannot provide more context, as this PR was made way before I joined the project. It's possible the PR only represents the outcome of other discussions but I can't justify that as I wasn't there 😅. What I was trying to say is that the table represents the logic present in the project for months now and this PR aims to fix a specific scenario preserving those rules.

> Since the decision was to proceed with this work for the patch release, I'll propose circling back to solve these problems in our next cycle.

Agree! 👍